src/blockchain_utilities/blockchain_and_pool.h

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2014-2020, The Monero Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice,
+// this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+// this list
+//    of conditions and the following disclaimer in the documentation and/or
+//    other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors
+// may be
+//    used to endorse or promote products derived from this software without
+//    specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include "cryptonote_core/blockchain.h"
+#include "cryptonote_core/tx_pool.h"
+
+struct BlockchainAndPool {
+  cryptonote::Blockchain blockchain;
+  cryptonote::tx_memory_pool tx_pool;
+
+  BlockchainAndPool() : blockchain(tx_pool), tx_pool(blockchain) {}
+};
```

Comment:
It seems strange to initialize an object with an uninitialized other object. Would this pattern be more intuitive?
```
struct BlockchainAndPool {
  cryptonote::Blockchain blockchain;
  cryptonote::tx_memory_pool tx_pool;

  BlockchainAndPool() : blockchain(), tx_pool(blockchain) {
    blockchain.set_pool(tx_pool);
  }
};
```

Reply:
I was following how `cryptonode::core` does it here:
https://github.com/monero-project/monero/blob/528cd729687fcc73e2c082225af203d5f67b3eb6/src/cryptonote_core/cryptonote_core.cpp#L228

@moneromooo-monero has proposed this pattern on IRC/Matrix

Maybe a problem with using a setter would be that if the setter is not called the object would be in an invalid state.
Another reason is that `Blockchain` and `tx_memory_pool` hold a reference to each other and those references need to be initialized in the constructor.

Reply:
I see. The original design is dubious but I guess yours is the best solution without more intrusive changes.